src/Validation/ValidationContext.php

```diff
@@ -0,0 +1,281 @@
+    /**
+     * @inheritdoc
+     */
+    public function getParentType(): ?TypeInterface
+    {
+        return $this->typeInfo->getParentType();
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function getFieldDefinition(): ?Field
+    {
+        return $this->typeInfo->getFieldDefinition();
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function getSchema(): SchemaInterface
+    {
+        return $this->schema;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function getArgument(): ?Argument
+    {
+        return $this->typeInfo->getArgument();
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function getDirective(): ?Directive
+    {
+        return $this->typeInfo->getDirective();
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function getFragment(string $name): ?FragmentDefinitionNode
+    {
+        if (empty($this->fragments)) {
+            $this->fragments = array_reduce($this->document->getDefinitions(), function ($fragments, $definition) {
+                if ($definition instanceof FragmentDefinitionNode) {
+                    $fragments[$definition->getNameValue()] = $definition;
+                }
+                return $fragments;
+            }, []);
+        }
+
+        return $this->fragments[$name] ?? null;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function getFragmentSpreads(SelectionSetNode $selectionSet): array
+    {
+        $spreads = $this->fragmentSpreads[(string)$selectionSet] ?? null;
+
+        if (null === $spreads) {
+            $spreads = [];
+
+            $setsToVisit = [$selectionSet];
+
+            while (!empty($setsToVisit)) {
+                /** @var SelectionSetNode $set */
+                $set = array_pop($setsToVisit);
+
+                /** @var FieldNode $selection */
+                foreach ($set->getSelections() as $selection) {
+                    if ($selection instanceof FragmentSpreadNode) {
+                        $spreads[] = $selection;
+                    } elseif ($selection->hasSelectionSet()) {
+                        $setsToVisit[] = $selection->getSelectionSet();
+                    }
+                }
+            }
+
+            $this->fragmentSpreads[(string)$selectionSet] = $spreads;
+        }
+
+        return $spreads;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function getRecursiveVariableUsages(OperationDefinitionNode $operation): array
+    {
+        $usages = $this->recursiveVariableUsages[(string)$operation] ?? null;
+
+        if (null === $usages) {
+            $usages    = $this->getVariableUsages($operation);
+            $fragments = $this->getRecursivelyReferencedFragments($operation);
+
+            foreach ($fragments as $fragment) {
+                // TODO: Figure out a more performance way to do this.
+                /** @noinspection SlowArrayOperationsInLoopInspection */
+                $usages = array_merge($usages, $this->getVariableUsages($fragment));
```

Comment:
Note that in order to merge 2 array deeply you need to use `array_merge_recursive`

Reply:
This is a flat array.

Reply:
ok :D